File: Leetcode/876. Middle of the Linked List/solution.cpp
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        if (!head) {
            return nullptr;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        while (true) {
            if (!fast->next) {
                return slow;
            }
            if (!fast->next->next) {
                return slow->next;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
    }
};

File: Leetcode/876. Middle of the Linked List/solution.rs
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() {
            return None;
        }
        let mut slow = head.clone();
        let mut fast = head.clone();
        loop {
            if fast.as_ref()?.next.is_none() {
                return slow;
            }
            if fast.as_ref()?.next.as_ref()?.next.is_none() {
                return slow.as_mut()?.next.take();
            }
            slow = slow?.next;
            fast = fast?.next?.next;
        }
    }
}
